Objects/tupleobject.c

```diff
@@ -76,11 +76,10 @@ _PyTuple_DebugMallocStats(FILE *out)
 #endif
 }
 
-PyObject *
-PyTuple_New(Py_ssize_t size)
+static PyObject *
+tuple_new_internal(Py_ssize_t size, int initialize)
```

Comment:
Uninitialized tuples must be initialized with Py_SIZE(op) = 0 since they are tracked by the GC (_PyObject_GC_TRACK call below). A GC collection can be triggered while the tuple is filled, whereas tupletraverse() uses Py_SIZE() to iterate on items. If Py_SIZE() isn't 0, tupletraverse() will likely crash.

I suggest the function name "tuple_new_impl".

Reply:
> Uninitialized tuples must be initialized with Py_SIZE(op) = 0 since they are tracked by the GC (_PyObject_GC_TRACK call below). A GC collection can be triggered while the tuple is filled, whereas tupletraverse() uses Py_SIZE() to iterate on items. If Py_SIZE() isn't 0, tupletraverse() will likely crash.

I don't like setting `Py_SIZE(op) = 0` and I see these alternatives:

1. if I understand correctly GC collection can be triggered only when some objects are Py_DECREFed, if that's true we can limit usage of proposed function only to the cases when it's quite obvious that there are no Py_DECREFs until tuple is filled (it's done this way in https://github.com/python/cpython/pull/8332), in the current patch there are just 2 not so obvious cases when `PyDict_Next()` is used
2. call `_PyObject_GC_TRACK()` in caller when tuple is filled as you proposed in https://github.com/python/cpython/pull/8332#discussion_r206582526

Reply:
> I suggest the function name "tuple_new_impl".

There is already this one:
https://github.com/python/cpython/blob/5105483acb3aca318304bed056dcfd7e188fe4b5/Objects/tupleobject.c#L696-L707